src/lib.rs

```diff
@@ -0,0 +1,320 @@
+// to increment the reference count (which will make it 2 in this particular case), and
+// mem::forgets the clone. The mem::forget prevents the reference count from being decremented when
+// we exit this function, so it will stay at 2 as long as we are in Rust code. Once the caller
+// drops its Arc, the reference count will go back down to 1, indicating the C code's copy.
+//
+// Unsafety:
+//
+// v must be a non-null pointer that resulted from previously calling `Arc::into_raw`.
+unsafe fn arc_with_incref_from_raw<T>(v: *const T) -> Arc<T> {
+    let r = Arc::from_raw(v);
+    let val = Arc::clone(&r);
+    mem::forget(r);
+    val
+}
+
+// Create a new rustls::ClientSession, and return it in the output parameter `out`.
+// If this returns an error code, the memory pointed to by `session_out` remains unchanged.
+// If this returns a non-error, the memory pointed to by `session_out` is modified to point
+// at a valid ClientSession. The caller now owns the ClientSession and must call
+// `rustls_client_session_free` when done with it.
+#[no_mangle]
+pub extern "C" fn rustls_client_session_new(
+    config: *const c_void,
+    hostname: *const c_char,
+    session_out: *mut *mut c_void,
+) -> CrustlsResult {
+    let hostname: &CStr = unsafe {
+        if hostname.is_null() {
+            eprintln!("rustls_client_session_new: hostname was NULL");
+            return CRUSTLS_ERROR;
+        }
+        CStr::from_ptr(hostname)
+    };
+    let config: Arc<ClientConfig> = unsafe {
+        match (config as *const ClientConfig).as_ref() {
+            Some(c) => arc_with_incref_from_raw(c),
+            None => {
+                eprintln!("rustls_client_session_new: config was NULL");
+                return CRUSTLS_ERROR;
+            }
+        }
+    };
+    let hostname: &str = match hostname.to_str() {
+        Ok(s) => s,
+        Err(e) => {
+            eprintln!("converting hostname to Rust &str: {}", e);
+            return CRUSTLS_ERROR;
+        }
+    };
+    let name_ref = match webpki::DNSNameRef::try_from_ascii_str(hostname) {
+        Ok(nr) => nr,
+        Err(e) => {
+            eprintln!(
+                "turning hostname '{}' into webpki::DNSNameRef: {}",
+                hostname, e
+            );
+            return CRUSTLS_ERROR;
+        }
+    };
+    let client = ClientSession::new(&config, name_ref);
+
+    // We've succeeded. Put the client on the heap, and transfer ownership
+    // to the caller. After this point, we must return CRUSTLS_OK so the
+    // caller knows it is responsible for this memory.
+    let b = Box::new(client);
+    unsafe {
+        *session_out = Box::into_raw(b) as *mut c_void;
+    }
+
+    return CRUSTLS_OK;
+}
+
+#[no_mangle]
+pub extern "C" fn rustls_client_session_wants_read(session: *const c_void) -> bool {
+    unsafe {
+        match (session as *const ClientSession).as_ref() {
+            Some(cs) => cs.wants_read(),
+            None => false,
```

Comment:
In what case would we see `None` here? if `session` were `NULL`?

Reply:
Yep, `as_ref` turns a raw pointer into an `Option<&T>`, which is None if the pointer is `NULL`.